Be much more efficient

Dr-Emann opened the pull request:
Try to remove as many extra allocations as possible. This also removes the dependency on the regex crate.

It's worth noting that this has some fairly breaking changes, and would therefore not be eligible to stay in the 1.0 stream.

Dr-Emann committed:
Be much more efficent.

Try to remove as many extra allocations as possible. This also removes
the dependency on the regex crate.

Dr-Emann committed:
Implement std::error::Error for HashIdsError

This has the side effect of implementing Debug, which means the Result
returned from the constructors can be `.unwrap()`ed

archer884 commented:
@Dr-Emann what's the idea rationale for using boxed slices and strs as opposed to just Vecs and Strings?

archer884 commented on src/lib.rs:
```diff
@@ -1,411 +1,333 @@
+        let mut buffer = String::new();
+        for number in numbers {
+          write!(buffer, "{:x}", number).unwrap();
+          result.push_str(&buffer[1..])
```
The second time through this for loop, you're going to be pushing a buffer that now contains _both_ of the numbers you've written to it. That's probably not good, right?

Dr-Emann committed:
clear buffer after each use

Dr-Emann commented:
Just saving a few bytes here and there (no capacity field). it's basically a zero cost operation to convert String->Box<str> or Vec<T> -> Box<[T]>, and it prevents accidental changes to the length.

Dr-Emann committed:
Changes for speed

Dr-Emann committed:
Fixes for padded hashes

Dr-Emann committed:
Remove check for empty numbers

Dr-Emann committed:
Better support some unicode.

Javascript uses UCS2, so it splits the salt/alphabet differently for
high characters

Dr-Emann committed:
Add failing test for null byte

Dr-Emann committed:
Add some documentation

Dr-Emann commented:
@charsyam when you get a chance, could you take a look at this?
